core/src/main/scala/org/bykn/bosatsu/Normalization.scala

```diff
@@ -0,0 +1,666 @@
+    val maxLambdaVar = None
+  }
+  case class Match(arg: NormalExpression,
+    branches: NonEmptyList[(NormalPattern, NormalExpression)])
+  extends NormalExpression {
+    val maxLambdaVar =
+      (arg.maxLambdaVar.toList ++ branches.toList.flatMap(_._2.maxLambdaVar))
+        .reduceLeftOption(Math.max)
+  }
+  case class LambdaVar(index: Int) extends NormalExpression {
+    val maxLambdaVar = Some(index)
+  }
+  /*
+   * It is reasonable to ask how you can define a lambda without an identifier
+   * for its argument in its expression. This is a benefit of de bruijn indexing.
+   * When all lambdas have exactly one argument you identify the var by how many
+   * lambdas out you have to travel.
+   *
+   * eg \x -> \y -> [y,x] would have normalization Lambda(Lambda(Apply(Apply(List, LambdaVar(1)), LambdaVar(0))))
+   *
+   * ref: https://en.wikipedia.org/wiki/De_Bruijn_index
+   */
+  case class Lambda(expr: NormalExpression) extends NormalExpression {
+    val maxLambdaVar = expr.maxLambdaVar.map(_ - 1)
+  }
+  case class Struct(enum: Int, args: List[NormalExpression]) extends NormalExpression {
+    val maxLambdaVar = args.flatMap(_.maxLambdaVar).reduceLeftOption(Math.max)
+  }
+  case class Literal(lit: Lit) extends NormalExpression {
+    val maxLambdaVar = None
+  }
+  case class Recursion(lambda: NormalExpression) extends NormalExpression {
+    val maxLambdaVar = lambda.maxLambdaVar
+  }
+}
+
+sealed abstract class NormalPattern {}
+
+object NormalPattern {
+  case object WildCard extends NormalPattern
+  case class Literal(toLit: Lit) extends NormalPattern
+  case class Var(name: Int) extends NormalPattern
+  /**
+   * Patterns like foo @ Some(_)
+   * @ binds tighter than |, so use ( ) with groups you want to bind
+   */
+  case class Named(name: Int, pat: NormalPattern) extends NormalPattern
+  case class ListPat(parts: List[Either[Option[Int], NormalPattern]]) extends NormalPattern
+  case class PositionalStruct(name: Option[Int], params: List[NormalPattern]) extends NormalPattern
+  case class Union(head: NormalPattern, rest: NonEmptyList[NormalPattern]) extends NormalPattern
+}
+
+object Normalization {
+  case class NormalExpressionTag(ne: NormalExpression, children: Set[NormalExpression])
+  type NormalizedPM = PackageMap.Typed[(Declaration, Normalization.NormalExpressionTag)]
+  type NormalizedPac = Package.Typed[(Declaration, Normalization.NormalExpressionTag)]
+
+  type PatternEnv = Map[Int, NormalExpression]
+
+  sealed trait PatternMatch[+A]
+  case class Matches[A](env: A) extends PatternMatch[A]
+  case object NoMatch extends PatternMatch[Nothing]
+  case object NotProvable extends PatternMatch[Nothing]
+
+  val noop: (NormalExpression, PatternEnv) => PatternMatch[PatternEnv] = { (_, _) => Matches(Map()) }
+  val neverMatch: (NormalExpression, PatternEnv) => PatternMatch[Nothing] = { (_, _) => NoMatch }
+
+  def structListAsList(ne: NormalExpression): PatternMatch[List[NormalExpression]] = {
```

Comment:
can you comment on the meaning of this? Is this just the analog of the Value type for checking the, representation of Lists?

Reply:
I think that's right. It's for when you have a `NormalExpression` and you're trying to see if it pattern matches to a list pattern, so you attempt to unwrap the `Struct` into a list. This only works if the normal expression is fully represented as a list (eg `Struct(1, List(a, Struct(1, List(b, Struct(0, List())))))`)

Note we only use this for when we have a `splice` followed by a pattern because we need to do that reversing of the tail trick